core/domain/topic_fetchers.py

```diff
@@ -455,3 +458,42 @@ def get_all_topic_rights():
         rights = get_topic_rights_from_model(model)
         topic_rights[rights.id] = rights
     return topic_rights
+
+
+def get_canonical_story_dicts(user_id, topic):
+    """Returns a list of canonical story dicts in the topic.
+
+    Args:
+        user_id: str. The ID of the user.
+        topic: Topic. The topic domain object.
+
+    Returns:
+        list(dict). A list of canonical story dicts in the given topic.
+    """
+    canonical_story_ids = topic.get_canonical_story_ids(
+        include_only_published=True)
+    canonical_story_summaries = [
+        story_fetchers.get_story_summary_by_id(
+            canonical_story_id) for canonical_story_id
+        in canonical_story_ids]
+    canonical_story_dicts = []
+    for story_summary in canonical_story_summaries:
+        all_nodes = story_fetchers.get_pending_and_all_nodes_in_story(
+            user_id, story_summary.id)['all_nodes']
+        pending_nodes = story_fetchers.get_pending_and_all_nodes_in_story(
+            user_id, story_summary.id)['pending_nodes']
```

Comment:
No, don't do this. Each of these does a `get` call. Just call this once in the loop, store it in a variable and use that. (Probably fine to loop the `get` since a topic ideally has only one, maybe 2 stories ).

Reply:
Done.